### PR TITLE
More API guidelines

### PR DESCRIPTION
**Pull request by cole-h**
I'm not totally told on this, but I think it might be a good idea. I dunno.

---

Based on #3.

### CODE REVIEW

**Comment by cole-h** [comment]
I dunno, I kinda like this now...

File: src/lib.rs [hunk 1]
```diff
@@ -1,168 +1,236 @@
...
+    fn from(value: String) -> Self {
+        Self(value)
+    }
+}
+
+/// A newtype wrapper around a [`String`] that represents a `nix.conf` setting
+/// value.
+#[derive(Clone, Eq, PartialEq, Ord, PartialOrd, Hash, Debug, Default)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub struct NixConfigValue(pub String);
```
Comment (Hoverbear) [inline src/lib.rs]: Is this always a string? Are booleans not a thing?

File: src/lib.rs [hunk 2]
```diff
@@ -1,168 +1,236 @@
...
+    }
+}
+
 /// A newtype wrapper around a [`HashMap`], where the key is the name of the Nix
 /// setting, and the value is the value of that setting. If the setting accepts
 /// a list of values, the value will be space delimited.
 #[derive(Clone, Eq, PartialEq, Debug, Default)]
-pub struct NixConfig(pub HashMap<String, String>);
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+#[cfg_attr(feature = "serde", serde(transparent))]
```
Comment (cole-h) [inline src/lib.rs]: I think this is the thing I'm most meh on... If we add new fields, we wouldn't be able to do this anymore (without implementing the `serde` stuff by hand. I could drop this attr, and the receipt over in `nix-installer` would then look like `nix_config: { settings: {  experimental-features = flakes, cores = 42 } }`, etc...

File: src/lib.rs (hunk 1 above)
Comment (cole-h) [reply src/lib.rs]: The underlying config file format in Nix is not typed, so I don't think it matters: https://github.com/NixOS/nix/blob/0079d2943702a7a7fbdd88c0f9a5ad677c334aa8/src/libutil/config.cc#L128

File: src/lib.rs [hunk 3]
```diff
@@ -1,168 +1,236 @@
...
-    let mut settings = NixConfig::new();
-
-    for line in contents.lines() {
-        let mut line = line;
-
-        // skip comments
-        if let Some(pos) = line.find('#') {
-            line = &line[..pos];
-        }
+    pub fn to_settings(&self) -> HashMap<NixConfigKey, NixConfigValue> {
```
Comment (Hoverbear) [inline src/lib.rs]: Since `NixConfig` already implements `Clone` and we offer `into_settings()` I'm not sure this is necessary API but that's fine.

**Comment by cole-h** [comment]
OK I think I like this now. Let's do a 0.1.0!

**Pull request merged**